dali/kernels/imgproc/convolution/convolution_gpu.h

```diff
@@ -0,0 +1,249 @@
+ *
+ * If the anchor is not provided, the window is centered over input:
+ * `window_anchor = window_size / 2`
+ *
+ * Only odd windows with centered anchor are now supported.
+ *
+ * Operation is done by using GEMM (General Matrix Multiplication) and generating the matrix
+ * from convolution window on the fly.
+ *
+ * CUTLASS allows to calculate GEMM: ``D = alpha * A * B + beta * C``.
+ * C and D can be the same matrix, giving us an equivalent of: ``D *= beta; D += alpha * A * B``
+ *
+ * To calculate the convolution, we modified the GEMM implementation so that it generates one of the
+ * input matrices on the fly based on the convolution window.
+ * Such a matrix can be generated by applying the convolution window to an identity matrix of the
+ * desired size.
+ *
+ * For the convolution we pass: alpha = `scale`, beta = 0 and D = C = `out`.
+ * We're intersted in the multiplication only, the C matrix is zeroed by the provided beta.
+ * A and B correspond to `in` and `window` (the order depends on whether it's an inner or an outer
+ * convolution). Effectively, we calculate ``D = alpha * A * B``.
+ */
+template <typename Out, typename In, typename W, int ndim, int axis, bool has_channels = true>
+struct ConvolutionGpu {
+  KernelRequirements Setup(KernelContext& ctx, const TensorListShape<ndim>& in_shape,
+                           const TensorListShape<1>& window_size) {
+    KernelRequirements req;
+    ScratchpadEstimator se;
+    DALI_ENFORCE(
+        in_shape.size() == window_size.size(),
+        make_string(
+            "Provided input shape and window sizes should have the same number of samples. Got: ",
+            in_shape.size(), " vs ", window_size.size(), "."));
+    int num_samples = in_shape.size();
+    for (int i = 0; i < num_samples; i++) {
+      int num_channels = has_channels ? in_shape[i][ndim - 1] : 1;
+      DALI_ENFORCE(
+          window_size[i][0] % 2 == 1,
+          make_string(
+              "Even or non-centered windows are not supported yet, got window with even length: ",
+              window_size, " for sample ", i, "."));
+
+      DALI_ENFORCE(window_size[i][0] * num_channels < kMaxWindowSize,
+                   make_string("Window is too big for sample ", i, ", got: ", window_size[i][0],
+                               ", expected at most: ", kMaxWindowSize / num_channels, "."));
+    }
+    se.add<W>(AllocType::Host, num_samples * kWindowCopyBufferSize);
+    se.add<W>(AllocType::GPU, num_samples * kWindowCopyBufferSize);
+    se.add<typename CutlassConv::SampleParams>(AllocType::GPU, num_samples);
+    req.scratch_sizes = se.sizes;
+    req.output_shapes.push_back(in_shape);
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorListView<StorageGPU, Out, ndim> out,
+           const TensorListView<StorageGPU, const In, ndim>& in,
+           const TensorListView<StorageCPU, const W, 1>& windows,
+           span<const int> window_anchors = {}, W scale = 1) {
```

Comment:
```suggestion
           span<const int> window_anchors = {}, float scale = 1) {
```
This would be useful if W is an integral type, so we can normalize.

Reply:
Done, both in CPU and GPU.